File: ProblemSet/464 · 整数排序 II/python/464 · 整数排序 II 快速排序 单向扫描.py
import random

class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.n = len(A)
        self.quickSort(A, 0, len(A) - 1)

    def quickSort(self, nums, left, right):
        if left < right:
            pos = self.partition(nums, left, right)
            self.quickSort(nums, left, pos - 1)
            self.quickSort(nums, pos + 1, right)

    
    def partition(self, nums, left, right):
        pivot = random.randint(0, right - left) + left
        nums[left], nums[pivot] = nums[pivot], nums[left]
        i = left
        for j in range(left + 1, right + 1):
            if nums[j] < nums[left]:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]
        nums[left], nums[i] = nums[i], nums[left]
        return i

File: ProblemSet/464 · 整数排序 II/java/464 · 整数排序 II 快速排序 单向扫描.java
import java.util.Random;

public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        quickSort(A, 0, A.length - 1);
    }
    
    private void quickSort(int[] nums, int left, int right) {
        if (left < right) {
            int pos = partition(nums, left, right);
            quickSort(nums, left, pos - 1);
            quickSort(nums, pos + 1, right);
        }
    } 

    private int partition(int[] nums, int left, int right) {
        int pivot = new Random().nextInt(right - left + 1) + left;
        swap(nums, left, pivot);
        int i = left;
        for (int j = left + 1; j <= right; ++j) {
            if (nums[j] < nums[left]) 
                swap(nums, ++i, j);
        }
        swap(nums, left, i);
        return i;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}